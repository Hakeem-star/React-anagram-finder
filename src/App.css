.App {
  text-align: center;
}



.inputSubmitContainer, .container {
  display: flex;
  flex-direction: column;
  align-items: center;
  width:100%;
}


h1 {
font-family: Arial, Helvetica, sans-serif;
font-weight: 400;
font-size: 40px;
}

.headerPortion{
  position: sticky;
  top:0;  
  display: flex;
  align-items: center;
  justify-content: center; 
  width:100%;
  background-color: rgb(0, 0, 0);
  border-bottom: 5px solid rgb(182, 182, 182);
  color:white;
}

h1{
  font-weight: 600;
}

.selector {
  color:rgb(0, 0, 0);
padding: 5px 3px 5px 30px;
  height: 100%;
  font-size: 25px;
  font-weight: 600;
  width: 220px;
border:0;
-webkit-appearance: none; 
background: url(http://cdn1.iconfinder.com/data/icons/cc_mono_icon_set/blacks/16x16/br_down.png) no-repeat right rgb(255, 255, 255);
background-position-x: 175px;
transition: 0.3s;
  border-radius: 20px 20px;

}
.selector:focus{
  outline: none;
}

.selector:hover {
  background-repeat: repeat-y;
  animation: selectorHoverArrow 1s infinite linear;

} 

@keyframes selectorHoverArrow {
  from {  background-position-y: 50%;}
  to {  background-position-y: 189%;}
}

.searchInput, .headerPortion, .selector{
margin-bottom: 50px;
}

.searchInput{
  width:100%;
  display:flex;
  flex-direction: row;
  justify-content: center;
  align-content: space-between;
  flex-wrap: wrap;
}

.searchInput > input[type="button"] {
width: 20%;
height: 35px;
background-color: rgb(0, 0, 0);
border-radius: 10%;
color:white;
font-weight: 800;
}

.searchTextInput {
  box-sizing: border-box;
  width: 100%;
font-size: 2em;
  background-color: rgb(255, 255, 255);
  border-color: rgb(182, 182, 182);
  border-width: 1px;
  margin:20px 5px 50px 5px;
  padding:8px;
  box-shadow:inset 2px 5px 8px 0px #e2e2e2;
  }
  


.inputName {
  display: flex;
  align-items: center;
  justify-content: center; 
  height: 16px;
  width:auto;
  background-color: rgb(255, 255, 255);
  padding:8px;
  border-color: rgb(182, 182, 182);
  border-width: 1px;
  border-style: inset;
font-weight: 600;
}

.inputNameGroup{
  display:flex;
  flex-direction: row;
  justify-content: center;
  align-items:stretch;
  width:80%;
  border:1px solid rgb(182, 182, 182);
  margin-bottom:40px;

}


table {
  border-collapse: collapse;
  width:80%;
}
table, td {
  border: 1px solid black;
}
th, td {
  padding: 15px;
  width:25%;
}

th {
  color:rgb(255, 255, 255);
  background-color: black;
  font-weight: 800;
  height: 35px;
  text-align: center;
}

.ResultsPageNumber{
  width:100%;
  display:flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.ResultsPageNumber > p{
  display:inline;
  margin:auto;
}

.next, .previous {
font-size:3em;
cursor: pointer;
}