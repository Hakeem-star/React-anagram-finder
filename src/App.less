@import "~antd/dist/antd.less";

#root {
  height: 100%;
}

.search {
  &:hover {
    border-top: 0;
  }

  &.ant-input-affix-wrapper-focused {
    border-top: 0;
  }
  .ant-input {
    font-size: 37px;
    padding-left: 40px;
  }
}

.main-content {
  padding-top: 60px;
  padding-bottom: 60px;
  display: inline-flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  width: 100%;
  text-align: left;

  &__threshold-container {
    max-width: 970px;
    width: 76%;
    .threshold-label {
      font-weight: 900;
      margin-bottom: 0;
      &__underline {
        width: 9.5%;
        background: #ced0d2;
        height: 1px;
        margin-top: 5px;
        margin-bottom: 20px;
      }
    }
  }
}

.footer {
  background-color: #e3e5e6;
}
