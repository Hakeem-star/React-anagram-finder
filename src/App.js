import React from 'react';
import celebAnagramFinder from './celebAnagramFinderAPICall';
import './App.css';


function NothingFound(props) {
  return (
    <h4>"Nothing was found!"</h4>
  )
}


class ResultsFound extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      currentResultsPage: 0,
      maxPages: 1,
    }
  }



  //Seperate the results into chunks, so we can have different pages for the results
  chunk(arr, chunckSize) {
    // debugger;
    let newChunk = [];
    let j = 0;
    for (let i = 0; i < arr.length; i += chunckSize) {
      !i ? j = 0 : j += chunckSize;
      newChunk.push(arr.slice(j, j + chunckSize))
    }

    return newChunk;
  }

  navigateResults(e) {
    if (e.target.textContent === " » ") {
      if (this.state.currentResultsPage < this.state.maxPages-1) {
        this.setState({ currentResultsPage: this.state.currentResultsPage + 1 })
      }
    } else if (e.target.textContent === " « "){
      if (this.state.currentResultsPage > 0) {
        this.setState({ currentResultsPage: this.state.currentResultsPage - 1 })
      }
    }
    console.dir(e.target.textContent);
  }
  componentDidMount() {
    //console.log( this.chunk(Object.keys(this.props.results),10));
    this.setState({ maxPages: this.chunk(Object.keys(this.props.results),10).length });
  }

  render() {

    let rows = (Array(Object.keys(this.props.results).length).fill("").map((value, ind) => {

      let anagramName = Object.keys(this.props.results)[ind]; //Thanksmo       
      let val = this.props.results[anagramName][0] //return the array associated with the anagram
      /*0: {Tom Hanks: "100%"} */
      let celebName = Object.keys(val)[0]; //Create an array based off the keys in the returned object. 
      //"Tom Hanks"
      let innerCurrent = val[celebName]
      return (
        <tr key={anagramName + celebName}>
          <td>{anagramName}</td>
          <td>{celebName}</td>
          <td>{innerCurrent}</td>
        </tr>
      )

    }));
    

    let resultsCounter = this.state.currentResultsPage > 0 ? 
    (this.chunk(rows, 10)[this.state.currentResultsPage].length) + (this.state.currentResultsPage *10) :
    this.chunk(rows, 10)[this.state.currentResultsPage].length

    return (
      <div className="container">
        <table>
          <thead>
            <tr>
              <th>Anagram</th>
              <th>Name</th>
              <th>Match Rate</th>
            </tr>
          </thead>
          <tbody>
            {this.chunk(rows, 10)[this.state.currentResultsPage]}
          </tbody>
        </table>
        <ResultsPageNumber prevResults={(e) => this.navigateResults(e)}
          current={resultsCounter}
          nextResults={(e) => this.navigateResults(e)} maximum={rows.length} />
      </div>
    )
  }
}



function ResultsPageNumber(props) {

  return (
    <div className="ResultsPageNumber">
      <p className="next" onClick={props.prevResults}> « </p>
      <p>Showing {props.current} of {props.maximum}</p>
      <p className="previous" onClick={props.nextResults}> » </p>
    </div>
  )
}

function ResultDivContain(props) {

  let results;

  if (Object.keys(props.results).length === 0) {
    results = <NothingFound />;
  } else {
    results = <ResultsFound
      page={props.resultsPage} results={props.results} />
  }

  return (
    <div className="container">
      {results}
    </div>
  )
}



function PageTitle(props) {

  return (
    <div className="headerPortion">
      <h1>{props.title + " Anagram Finder"}</h1>
    </div>
  )
}

function SearchSelector(props) {

  return (
    <div>
      <div >
        <select className="selector" onChange={props.handleClick} name="anagramQuerySelect" id="anagramQuerySelect">
          <option className="celebrities" value="Celebrities">Celebrities</option>
          <option className="general" value="General">General</option>
        </select>
      </div>
    </div>
  )
}


function SearchInput(props) {
  let placeholderText;
  props.placeholder === "General" ? placeholderText = "This feature is coming soon" : placeholderText = "Thanksmo,TRUMCDONALD,parsi/hl itno";

  return (
    <div className="searchInput">
      <input
        className="searchTextInput"
        onKeyPress={props.submitted}
        onChange={props.searchTerm}
        type="text"
        defaultValue={placeholderText}>
      </input>
      <input onClick={props.submitted} type="button" value="Search" />
    </div>
  )
}


class App extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      anagramType: PageTitle.defaultProps.title,
      SearchInputName: PageTitle.defaultProps.title,
      searchTerm: "",
      results: "",
      resultsIndex: 0,

    }
  }

  searchSelected(e) {
    //console.dir(e.target)
    this.setState({ anagramType: e.target.value, SearchInputName: e.target.value });
  }

  async localCelebAnagramFinder(val) {
    let test = await celebAnagramFinder(val);
    this.setState({ results: test });
  }

  searchSubmitted(e) {
    if (e.key === "Enter" || e.target.type === "button") {
      this.localCelebAnagramFinder(document.querySelectorAll(".searchTextInput")[0].value);
    }
  }

  render() {
    let rows;
    if (typeof (this.state.results) === "string") {
      rows = [];
    } else {
      rows = <ResultDivContain
        results={this.state.results} prevResults={this.props.gotoPrevResults}
        nextResults={this.props.gotoNextResults} current={this.state.resultsIndex} />
    }


    return (
      <div>

        <div>
          <PageTitle title={this.state.anagramType} />

          <div className="inputSubmitContainer">
            <SearchSelector handleClick={(e) => this.searchSelected(e)} />
            <SearchInput placeholder={this.state.anagramType} submitted={(e) => this.searchSubmitted(e)} />
          </div>
          <div>
            {rows}
          </div>
        </div>
      </div>
    );
  }
}


PageTitle.defaultProps = {
  title: "Celebrity",
  SearchInputName: "Name",
}

export default App;
