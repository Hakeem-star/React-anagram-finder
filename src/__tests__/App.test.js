import React from "react";
import App from "../App";
import { waitFor, render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

jest.mock("../utils/fetchCelebData");

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
Object.defineProperty(window.HTMLElement.prototype, "scrollIntoView", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({})),
});

function searchAndEnter(input, searchterm) {
  fireEvent.input(input, { target: { value: searchterm } });
  fireEvent.keyDown(input, { key: "Enter", keyCode: 13 });
}
// beforeEach()

describe("<App />", () => {
  test("renders without crashing", () => {
    render(<App />);
  });

  test("Displays a message when there's no input", async () => {
    const { findAllByText, getByPlaceholderText } = render(<App />);
    const input = getByPlaceholderText("input anagram");
    searchAndEnter(input, "");
    const noAnagram = await findAllByText("Please type in an anagram");
    expect(noAnagram.length).toBe(1);
  });

  test("Displays a results table when a result is found", async () => {
    const { findAllByText, getByPlaceholderText } = render(<App />);
    const input = getByPlaceholderText("input anagram");
    fireEvent.input(input, { target: { value: "trump" } });
    fireEvent.keyDown(input, { key: "Enter", keyCode: 13 });
    expect(await findAllByText(/donald/i));
  });

  describe("Anagram Type toggle", () => {
    test("It changes the anagrams function based on what you pick on the header", async () => {
      const { getByText, findAllByText, getByPlaceholderText } = render(
        <App />
      );
      const celebsTab = getByText("Celebs");
      const generalsTab = getByText("General");
      const input = getByPlaceholderText("input anagram");
      searchAndEnter(input, "trump");
      let findTrump = await findAllByText(/trump/i);
      expect(findTrump.length).toBe(3);
      fireEvent.click(generalsTab, { button: 1 });
      searchAndEnter(input, "trumpet");
      const findTrumpet = await findAllByText(/trumpet/i);
      findTrump = await findAllByText("trump");
      expect(findTrumpet.length).toBe(4);
      expect(findTrump.length).toBe(1);
    });
  });

  test("It highlights a historical search item when a previous search is repeated ", async () => {
    const { findAllByText, getByPlaceholderText } = render(<App />);
    const input = getByPlaceholderText("input anagram");

    searchAndEnter(input, "trump//");
    await findAllByText("trump//");
    searchAndEnter(input, "test");
    await findAllByText("test");
    searchAndEnter(input, "trump//");
    const trumpResult = await findAllByText("trump//");
    expect(trumpResult[0].parentElement.parentElement).toHaveClass("--active");
    searchAndEnter(input, "test");
    await waitFor(() => {}, { timeout: 50 });
    expect(trumpResult[0].parentElement.parentElement).not.toHaveClass(
      "--active"
    );
  });

  test("Displays 'No results found' when a result is not found", async () => {
    const { findAllByText, getByPlaceholderText } = render(<App />);
    const input = getByPlaceholderText("input anagram");
    searchAndEnter(input, "//.");
    expect(await findAllByText(/No results found/i));
  });

  test("clicking on previous history searches for historical anagram", async () => {
    const { findAllByText, getByPlaceholderText } = render(<App />);
    const input = getByPlaceholderText("input anagram");
    //Input element is not included in findby results, so visible counts will be 1 less

    //Search for trump
    searchAndEnter(input, "trump");
    //wait for result
    let findDonald = await findAllByText(/trump/i);
    expect(findDonald.length).toBe(3);

    //Search for test
    searchAndEnter(input, "test");
    //wait for result
    const testResult = await findAllByText(/test/i);
    findDonald = await findAllByText(/trump/i);
    //Trump from history still visible
    expect(findDonald.length).toBe(1);

    expect(testResult.length).toBe(3);

    //Open the history toggle
    const expandHistoryToggle = document.querySelector(
      ".sider-container__icon-container__icon.history-icon"
    );
    fireEvent.click(expandHistoryToggle, { button: 1 });

    //Click on the trump history
    const trumpHistoryButton = document.querySelector(".result-0 button");
    fireEvent.click(trumpHistoryButton, { button: 1 });
    const secondTestResult = await findAllByText("test");
    findDonald = await findAllByText(/trump/i);
    //wait for result
    // expect(secondTestResult.length).toBe(1);
    expect(findDonald.length).toBe(3);
  });
});
