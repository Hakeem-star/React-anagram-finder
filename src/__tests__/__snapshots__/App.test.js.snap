// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches the snapshot 1`] = `
<div>
  <div>
    <div
      className="headerPortion"
    >
      <h1>
        Celebrity Anagram Finder
      </h1>
    </div>
    <div
      className="inputSubmitContainer"
    >
      <div>
        <select
          className="selector"
          id="anagramQuerySelect"
          name="anagramQuerySelect"
          onChange={[Function]}
        >
          <option
            className="celebrities"
            value="Celebrities"
          >
            Celebrities
          </option>
          <option
            className="general"
            value="General"
          >
            General
          </option>
        </select>
      </div>
      <div
        className="searchInput"
      >
        <input
          className="searchTextInput"
          defaultValue="Thanksmo,TRUMCDONALD,parsi/hl itno"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
        />
        <input
          onClick={[Function]}
          type="button"
          value="Search"
        />
      </div>
    </div>
    <div
      className="container"
    >
      <div>
        <p
          id="historyLabel"
        >
          Previous Searches
        </p>
        <div
          id="historyLabelArrows"
        />
      </div>
      <div
        className="resultsContainer"
      />
    </div>
    <div
      className="container"
    >
      <div
        className="lds-dual-ring"
        style={
          Object {
            "display": "none",
          }
        }
      />
    </div>
    <div />
  </div>
</div>
`;
